add yanking guidance to the readme faq (#102636)
JuliaRegistries/General: README.md

Early merges should still be discussed on the [#pkg-registration](https://julialang.slack.com/archives/C6M4DQA5P) slack channel before merging.
Registry maintainers are discouraged from merging packages they directly *or* indirectly maintain as mentioned in [CONTRIBUTING.md](https://github.com/JuliaRegistries/General/blob/master/CONTRIBUTING.md#other-ways-to-help).

### When is yanking a release appropriate?

Releases that have already merged can be marked as "yanked" by adding `yanked = true` under the release entry in `Versions.toml`. This will mean 
that Pkg will no longer consider them to exist and not resolve those versions, but that version cannot be re-registed in General. Yanking should 
only be used when either:
1) adding a standard patch release will not resolve the issue i.e. say `v2.10.0` was released using a feature not on julia `v1.6` but the compat entry for julia
was not raised in the release. In this case releasing a `v2.10.1` with the corrected julia compat would not solve the issue as on julia v1.6 Pkg would still
resolve the broken `v2.10.0`, and as a minor bump, reverting the code changes would not be valid in a patch bump.

> 3) there are security issues with the release
> 

Author: ```suggestion
2) there are security issues with the release

If yanking is urgent, open a PR and raise it on the `#pkg-registration` slack channel
```

Early merges should still be discussed on the [#pkg-registration](https://julialang.slack.com/archives/C6M4DQA5P) slack channel before merging.
Registry maintainers are discouraged from merging packages they directly *or* indirectly maintain as mentioned in [CONTRIBUTING.md](https://github.com/JuliaRegistries/General/blob/master/CONTRIBUTING.md#other-ways-to-help).

### When is yanking a release appropriate?

Releases that have already merged can be marked as "yanked" by adding `yanked = true` under the release entry in `Versions.toml`. This will mean 
that Pkg will no longer consider them to exist and not resolve those versions, but that version cannot be re-registed in General. Yanking should 
only be used when either:
1) adding a standard patch release will not resolve the issue i.e. say `v2.10.0` was released using a feature not on julia `v1.6` but the compat entry for julia
was not raised in the release. In this case releasing a `v2.10.1` with the corrected julia compat would not solve the issue as on julia v1.6 Pkg would still
resolve the broken `v2.10.0`, and as a minor bump, reverting the code changes would not be valid in a patch bump.

> 2) there are security issues with the release

Reviewer: I would list this first.
And I would say more along the lines of:

> The primary reason yanking exists is to make a release impossible to install.
This is intended to be used for circumstances when **having this release installed poses a risk to security or safety**.
It is not intended for avoiding people installing releases with bugs, for that use a patch release -- potentially after doing nothing more than revert . All patch releases prior to the most recent have known bugs by definition.

> There is however, a special category of bugged releases that can not be resolved by having a patch release. These also need to be resolved by yanking. That special category is when the compat bounds have been set too wide. [Insert your example here]

Author: sounds good. updated

> Releases that have already merged can be marked as "yanked" by adding `yanked = true` under the release entry in `Versions.toml`. This will mean 
> that Pkg will no longer consider them to exist and not resolve those versions, but that version cannot be re-registed in General. Yanking should 
> only be used when either:
> 1) adding a standard patch release will not resolve the issue i.e. say `v2.10.0` was released using a feature not on julia `v1.6` but the compat entry for julia
> was not raised in the release. In this case releasing a `v2.10.1` with the corrected julia compat would not solve the issue as on julia v1.6 Pkg would still
> resolve the broken `v2.10.0`, and as a minor bump, reverting the code changes would not be valid in a patch bump.
> 2) there are security issues with the release

Author: ```suggestion
Releases that have already merged can be marked as "yanked" by adding `yanked = true` under the release entry in `Versions.toml`. This will mean 
that Pkg will no longer consider them to exist and not resolve those versions, but that version cannot be re-registed in General. 

The primary reason yanking exists is to make a release impossible to install.
This is intended to be used for circumstances when having this release installed poses a risk to security or safety.
It is not intended for avoiding people installing releases with bugs, for that use a patch release -- potentially after doing nothing more than revert. All patch releases prior to the most recent have known bugs by definition.

There is however, a special category of bugged releases that can not be resolved by having a patch release. These also need to be resolved by yanking. That special category is when the compat bounds have been set too wide.    i.e. say `v2.10.0` was released using a feature not on julia `v1.6` but the compat entry for julia
was not raised in the release. In this case releasing a `v2.10.1` with the corrected julia compat would not solve the issue as on julia v1.6 Pkg would still
resolve the broken `v2.10.0`, and as a minor bump, reverting the code changes would not be valid in a patch bump.
```

### Who can approve an early merge?

Any registry maintainer may merge a package registration PR early.
Early merges should still be discussed on the [#pkg-registration](https://julialang.slack.com/archives/C6M4DQA5P) slack channel before merging.
Registry maintainers are discouraged from merging packages they directly *or* indirectly maintain as mentioned in [CONTRIBUTING.md](https://github.com/JuliaRegistries/General/blob/master/CONTRIBUTING.md#other-ways-to-help).

### When is yanking a release appropriate?

Releases that have already merged can be marked as "yanked" by adding `yanked = true` under the release entry in `Versions.toml`. This will mean 
that Pkg will no longer consider them to exist and not resolve those versions, but that version cannot be re-registed in General. 

The primary reason yanking exists is to make a release impossible to install.

> This is intended to be used for circumstances when having this release installed poses a risk to security or safety.

Reviewer 2: How is security or safety defined?

> It is not intended for avoiding people installing releases with bugs, for that use a patch release -- potentially after doing nothing more than revert. All patch releases prior to the most recent have known bugs by definition.

Reviewer 2: While it is true that all patches prior have bugs, the goal of a patch release is to fix a bug. I agree that it would be weird to yank all prior variants of that release.

What should we do with a patch release that introduces a new bug? (And who determines criticalness?) The yank that has caused these recent discussions is was an atomic release (one that only changed one thing) and that introduced (from the kinds of usage the packages sees) a severe bug.

A new patch release reverting the change was made and the version was yanked. Modulo the judgement of the maintainer on severity that seems to be the exact purpose of yanking?

Reviewer 3: > What should we do with a patch release that introduces a new bug?

Ask maintainers to make a new patch release.

> Modulo the judgement of the maintainer on severity that seems to be the exact purpose of yanking?

A severe package/usage bug is not generally a security vulnerability. A package merely not working correctly is different to that package causing security vulnerabilities down the road for users of that package.

If there is some urgent need for the package to be registered (for example, an upcoming talk featuring the package or other external deadline), the waiting period can be expedited by request made in the [#pkg-registration](https://julialang.slack.com/archives/C6M4DQA5P) channel on the 
public Julia Language Slack, or directly on the PR itself. This request should include a link to the registration PR and why an expedited merge is needed. In this case, a registry maintainer (see below) can manually review the package (often with stricter standards than the typical package faces, given the lack of bandwidth to review all packages thoroughly).

### Who can approve an early merge?

Any registry maintainer may merge a package registration PR early.
Early merges should still be discussed on the [#pkg-registration](https://julialang.slack.com/archives/C6M4DQA5P) slack channel before merging.
Registry maintainers are discouraged from merging packages they directly *or* indirectly maintain as mentioned in [CONTRIBUTING.md](https://github.com/JuliaRegistries/General/blob/master/CONTRIBUTING.md#other-ways-to-help).

### When is yanking a release appropriate?

> Releases that have already merged can be marked as "yanked" by adding `yanked = true` under the release entry in `Versions.toml`. This will mean
> that Pkg will no longer consider them to exist and not resolve those versions, but that version cannot be re-registed in General.

Reviewer 4: ```suggestion
Releases that have already merged can be marked as "yanked" by adding `yanked = true` under the release entry in `Versions.toml`
([example](https://github.com/JuliaRegistries/General/pull/94858/files)). This will mean that Pkg will no
longer consider them to exist and not resolve those versions, but that version cannot be re-registed in General.
```

> The primary reason yanking exists is to make a release impossible to install. This is intended to be used for circumstances when having this

Reviewer 4: ```suggestion
The reason yanking exists is to make a release impossible to install. This is intended to be used for circumstances when having this
```
Are there any other reasons?

Reviewer: Yes, the second reason is because the package manager generally will chose to use an older patch release if that is what is required for compatibility with something else.
Which is a problem if the bug in the older patch release is incorrect compat bounds.

(I have argued that for this reason Pkg shouldn't actually install older patch releases without a special flag, but people disagreed, so yanking is often required when your bug is with compat bounds)

Reviewer 4: I think of the compat reason as a subset of "make a release impossible to install."

Reviewer 4: Though I guess the "primary" here refers to the following sentence, not the sentence it is a part of?

Reviewer 4: ```suggestion
The reason yanking exists is to make a release impossible to install. This is primary intended to be used for circumstances when having this
```

Reviewer: we should add this, plus a footote aon "install" saying

```
To be precise it will still be possible to install if using the same manifest.
However any new manifest, such as those created by `Pkg.test` or `] up SomeUnrelatedPackage`, or `resolve` after changing julia version, will fail.
```

This will address @MasonProtter 's comment.

release installed poses a risk to security or safety. It is not intended for avoiding people installing releases with bugs, for that use a

> patch release -- potentially after doing nothing more than revert. All patch releases prior to the most recent have known bugs by definition.

Reviewer 3: ```suggestion
patch release -- potentially after doing nothing more than `git revert`  in the package repo. All patch releases prior to the most recent have known bugs by definition.
```

There is however, a special category of bugged releases that can not be resolved by having a patch release. These also need to be resolved by
yanking. That special category is when the compat bounds have been set too wide. i.e. say `v2.10.0` was released using a feature not on julia
`v1.6` but the compat entry for julia was not raised in the release. In this case releasing a `v2.10.1` with the corrected julia compat would
not solve the issue as on julia v1.6 Pkg would still resolve the broken `v2.10.0`, and as a minor bump, reverting the code changes would not
be valid in a patch bump.

> If yanking is urgent, open a PR and raise it on the `#pkg-registration` slack channel

Reviewer 4: ```suggestion
If yanking is urgent, open a PR and raise it on the `#pkg-registration` [slack channel](https://julialang.org/slack/)
```
